appscale/tools/appcontroller_client.py

```diff
@@ -227,6 +227,42 @@ def start_roles_on_nodes(self, roles_to_nodes):
     return self.run_with_timeout(self.DEFAULT_TIMEOUT, "Error", self.DEFAULT_NUM_RETRIES,
       self.server.start_roles_on_nodes, roles_to_nodes, self.secret)
 
+  def is_appscale_terminated(self):
+    """Queries the AppController to see if the system has been terminated.
+
+    Returns:
+      A boolean indicating whether appscale has finished running terminate
+        on all nodes.
+    """
+    return self.run_with_timeout(self.DEFAULT_TIMEOUT,
+                                 "Error: Client Timed Out",
+                                 self.DEFAULT_NUM_RETRIES,
+                                 self.server.is_appscale_terminated, self.secret)
+
+  def run_terminate(self, clean):
+    """Tells the AppController to terminate AppScale on the deployment.
+
+    Args:
+      clean: A boolean indicating whether the clean parameter should be
+        passed to terminate.rb.
+    Returns:
+      The result of executing the SOAP call on the remote AppController.
+    """
+    return self.server.run_terminate(clean, self.secret)
+
+  def receive_server_message(self):
```

Comment:
I'm not sure this pattern is a good idea. Imagine a case where the request gets through to the server, the server clears the messages in the queue, but a connection issue causes a timeout on the client side. The message would essentially be lost, and the client will keep checking for the missing message.

My suggestion would be for `run_terminate` to return a unique string that associates messages with a given request. You can then add a function to get the status of a particular request instead of a generic function to get all waiting messages.